File: frontend/src/api/external-asset-data/os-names-data-source-handler.ts
import proj4 from "proj4";

import { Feature } from "geojson";
import { DataSourceHandler } from "./data-source-handler";
import { AssetSpecification } from "@/hooks/queries/dataset-utils";

proj4.defs(
  "EPSG:27700",
  "+proj=tmerc +lat_0=49 +lon_0=-2 +k=0.9996012717 " +
    "+x_0=400000 +y_0=-100000 +ellps=airy " +
    "+towgs84=446.448,-125.157,542.06,0.1502,0.2470,0.8421,-20.4894 " +
    "+units=m +no_defs",
);

export class OsNamesDataSourceHandler extends DataSourceHandler {
  /**
   * A function to build a list of OS Names URLs.
   *
   * @param assetSpecification the specification of the asset
   * @returns
   */
  public buildUrlsForDataSource(
    assetSpecification: AssetSpecification,
  ): string[] {
    return [
      `/transparent-proxy/os-names/search/names/v1/find?query=Wight&fq=BBOX:${this.locator}&fq=LOCAL_TYPE:${assetSpecification.description}`,
    ];
  }

  public async fetchDataForAssetSpecification(
    assetSpecification: AssetSpecification,
    url: string,
  ): Promise<Feature[]> {
    const response = await this.fetchFromUrl(url);
    return response.results ? this.mapResponseToFeatures(response) : [];
  }

  private mapResponseToFeatures(response): Feature[] {
    return response.results.map((r) => {
      const data = r.GAZETTEER_ENTRY;
      return {
        id: data.ID,
        type: "Feature",
        geometry: {
          type: "Point",
          coordinates: this.mapBngCoordinatesToLatitudeLongitude(
            data.GEOMETRY_X,
            data.GEOMETRY_Y,
          ),
        },
        properties: {
          name: data.NAME2 ?? data.NAME1,
        },
      };
    });
  }

  private mapBngCoordinatesToLatitudeLongitude(
    easting: any,
    northing: any,
  ): number[] {
    return proj4("EPSG:27700", "EPSG:4326", [easting, northing]);
  }
}

File: backend/vista-python-api/src/api/repository/external/handlers/os_names_data_source_handler.py
"""Handler for the OS Names data source."""

from functools import reduce

from django.conf import settings

from api.repository.external.external_asset_mapper import ExternalAssetMapper

from .base import DataSourceHandler


class OsNamesDataSourceHandler(DataSourceHandler):
    """Handler for the OS Names data source."""

    def build_urls_for_data_source(self, asset_specification):
        """Build the URLs for fetching the data per the specification given."""
        return [
            f"https://api.os.uk/search/names/v1/find?key={settings.OS_NAMES_API_KEY}\
            &query=Wight&fq=BBOX:{self.locator}&fq=LOCAL_TYPE:{asset_specification['description'][0]}"
        ]

    async def fetch_data_for_asset_specification(self, asset_specification, url):
        """Fetch the OS Names data per the specification given."""
        data = await self.fetch_from_url_with_retry(url)
        simplified_data = reduce(
            self._merge_entries_with_same_name_at_same_location(set()), data["results"], []
        )
        return [
            ExternalAssetMapper.map_from_os_names(result, asset_specification)
            for result in simplified_data
        ]

    def _merge_entries_with_same_name_at_same_location(self, seen):
        def reduce(entries, next_entry):
            entry = next_entry["GAZETTEER_ENTRY"]
            if not entry["NAME1"]:
                entries.append(entry)
            else:
                location = f"{entry['NAME1']}_{entry['GEOMETRY_X']}_{entry['GEOMETRY_Y']}"
                if location in seen:
                    return entries
                seen.add(location)
                entries.append(entry)
            return entries

        return reduce
